src/cgroups/v2/systemd_manager.rs

```diff
@@ -0,0 +1,304 @@
+    fn get_unit_name(&self) -> String {
+        // By default we create a scope unless specified explicitly.
+        if !self.cgroups_path.name.ends_with(".slice") {
+            return format!(
+                "{}-{}.scope",
+                self.cgroups_path.scope, self.cgroups_path.name
+            );
+        }
+        self.cgroups_path.name.clone()
+    }
+
+    // systemd represents slice hierarchy using `-`, so we need to follow suit when
+    // generating the path of slice. For example, 'test-a-b.slice' becomes
+    // '/test.slice/test-a.slice/test-a-b.slice'.
+    fn expand_slice(&self, slice: String) -> Result<PathBuf> {
+        let suffix = ".slice";
+        if slice.len() <= suffix.len() || !slice.ends_with(suffix) {
+            bail!("invalid slice name: {}", slice);
+        }
+        if slice.contains('/') {
+            bail!("invalid slice name: {}", slice);
+        }
+        let mut path = "".to_owned();
+        let mut prefix = "".to_owned();
+        let slice_name = slice.trim_end_matches(suffix);
+        // if input was -.slice, we should just return root now
+        if slice_name == "-" {
+            return Ok(Path::new("/").to_path_buf());
+        }
+        for component in slice_name.split('-') {
+            if component.is_empty() {
+                anyhow!("Invalid slice name: {}", slice);
+            }
+            // Append the component to the path and to the prefix.
+            path = format!("{}/{}{}{}", path, prefix, component, suffix);
+            prefix = format!("{}{}-", prefix, component);
+        }
+        Ok(Path::new(&path).to_path_buf())
+    }
+
+    // get_cgroups_path generates a cgroups path from the one provided by the user via cgroupsPath.
+    // an example of the final path: "/machine.slice/docker-foo.scope"
+    fn get_cgroups_path(&self) -> Result<PathBuf> {
+        // the root slice is under 'machine.slice'.
+        let mut slice = Path::new("/machine.slice").to_path_buf();
+        // if the user provided a '.slice' (as in a branch of a tree)
+        // we need to "unpack it".
+        if !self.cgroups_path.parent.is_empty() {
+            slice = self.expand_slice(self.cgroups_path.parent.clone())?;
+        }
+        let unit_name = self.get_unit_name();
+        let cgroups_path = slice.join(unit_name);
+        Ok(cgroups_path)
+    }
+
+    /// create_unified_cgroup verifies sure that *each level* in the downward path from the root cgroup
+    /// down to the cgroup_path provided by the user is a valid cgroup hierarchy,
+    /// containing the attached controllers and that it contains the container pid.
+    fn create_unified_cgroup(&self, pid: Pid) -> Result<PathBuf> {
+        let cgroups_path = self.get_cgroups_path()?;
+        let full_path = self.root_path.join_absolute_path(&cgroups_path)?;
+        let controllers: Vec<String> = self
+            .get_available_controllers(common::DEFAULT_CGROUP_ROOT)?
```

Comment:
Use the root path instead, in case the cgroup fs is not mounted in the default location

Reply:
Sounds good, thanks!